library/Booster/Pattern/ApplyEquations.hs

```diff
@@ -688,27 +688,64 @@ simplifyConstraints doTracing def mbApi cache ps =
     runEquationT doTracing def mbApi cache $ mapM (simplifyConstraint' True) ps
 
 -- version for internal nested evaluation
-simplifyConstraint' :: MonadLoggerIO io => Bool -> Predicate -> EquationT io Predicate
+simplifyConstraint' :: MonadLoggerIO io => Bool -> Term -> EquationT io Term
 -- We are assuming all predicates are of the form 'true \equals P' and
 -- evaluating them using simplifyBool if they are concrete.
 -- Non-concrete \equals predicates are simplified using evaluateTerm.
-simplifyConstraint' recurse = \case
-    Predicate t@(Term TermAttributes{canBeEvaluated} _)
+simplifyConstraint' recurseIntoEvalBool = \case
+    t@(Term TermAttributes{canBeEvaluated} _)
         | isConcrete t && canBeEvaluated -> do
             mbApi <- (.llvmApi) <$> getConfig
             case mbApi of
                 Just api ->
-                    if simplifyBool api t
-                        then pure $ Predicate TrueBool
-                        else pure $ Predicate FalseBool
-                Nothing ->
-                    Predicate <$> if recurse then evalBool t else pure t
-        | otherwise ->
-            Predicate <$> if recurse then evalBool t else pure t
+                    pure $
+                        if simplifyBool api t
+                            then TrueBool
+                            else FalseBool
+                Nothing -> if recurseIntoEvalBool then evalBool t else pure t
+    NotBool x -> negateBool <$> recursion x
+    EqualsK (KSeq _ l) (KSeq _ r) -> evalEqualsK l r
+    NEqualsK (KSeq _ l) (KSeq _ r) -> negateBool <$> evalEqualsK l r
+    t -> if recurseIntoEvalBool then evalBool t else pure t
   where
     evalBool :: MonadLoggerIO io => Term -> EquationT io Term
     evalBool t = do
         prior <- getState -- save prior state so we can revert
         result <- iterateEquations 100 TopDown PreferFunctions t
         EquationT $ lift $ lift $ put prior
         pure result
+
+    recursion = simplifyConstraint' recurseIntoEvalBool
+
+    negateBool TrueBool = FalseBool
+    negateBool FalseBool = TrueBool
+    negateBool x = NotBool x
+
+    foldAndBool [] = TrueBool
+    foldAndBool [x] = x
+    foldAndBool (FalseBool : _) = FalseBool
+    foldAndBool (TrueBool : xs) = foldAndBool xs
+    foldAndBool (x : xs) = AndBool x $ foldAndBool xs
+
+    evalEqualsK l@(SymbolApplication sL _ argsL) r@(SymbolApplication sR _ argsR)
+        | isConstructorSymbol sL && isConstructorSymbol sR =
+            if sL == sR
+                then foldAndBool <$> zipWithM evalEqualsK argsL argsR
+                else pure FalseBool
+        | otherwise =
+            (if recurseIntoEvalBool then evalBool else pure) $
+                EqualsK (KSeq (sortOfTerm l) l) (KSeq (sortOfTerm r) r)
```

Comment:
I am not sure about this recursion, would it not go into a loop if `recurseIntoEvalBool` was true?
Going like this:
1. call to `evalBool` with argument `EqualsK (KSeq s1 (f x)) (KSeq s2 (g y))` with functions `f` and `g`
2. line 707 calls `evalEqualsK (f x) (g y)`
3. line 736 calls `evalBool (EqualsK (KSeq s1' (f x)) (KSeq s2' (g y)))` (arriving at step 1)

I think we should at least not recurse in this case here. The second case in line 751 appears similar, we are reproducing the original call that led us to this case.